core/include/gnuradio-4.0/Block.hpp

```diff
@@ -1311,14 +1311,33 @@ class Block : public lifecycle::StateMachine<Derived>, //
     }
 
     template<typename TIn, typename TOut>
-    work::Status
-    invokeProcessBulk(TIn &inputSpans, TOut &outputSpans) {
-        // cannot use std::apply because it requires tuple_cat(inputSpans, writersTuple). The latter doesn't work because writersTuple isn't copyable.
+    gr::work::Status invokeProcessBulk(TIn &inputReaderTuple, TOut &outputReaderTuple) {
+        auto tempInputSpanStorage = std::apply([]<typename... PortReader>(PortReader&... args) {
+            return std::tuple{(gr::meta::array_or_vector_type<PortReader> ? std::span{args.data(), args.size()} : args)...};
+        }, inputReaderTuple);
+
+        auto tempOutputSpanStorage = std::apply([]<typename... PortReader>(PortReader&... args)  {
+            return std::tuple{(gr::meta::array_or_vector_type<PortReader> ? std::span{args.data(), args.size()} : args)...};
+        }, outputReaderTuple);
+
+        auto refToSpan = []<typename T, typename U>(T&& original, U&& temporary) -> decltype(auto) {
+            if constexpr (gr::meta::array_or_vector_type<std::decay_t<T>>) {
+                return std::forward<U>(temporary);
+            } else {
+                return std::forward<T>(original);
+            }
+        };
+
         return [&]<std::size_t... InIdx, std::size_t... OutIdx>(std::index_sequence<InIdx...>, std::index_sequence<OutIdx...>) {
-            return self().processBulk(std::get<InIdx>(inputSpans)..., std::get<OutIdx>(outputSpans)...);
-        }(std::make_index_sequence<std::tuple_size_v<std::remove_cvref_t<TIn>>>(), std::make_index_sequence<std::tuple_size_v<std::remove_cvref_t<TOut>>>());
+            return self().processBulk(
+                    refToSpan(std::get<InIdx>(inputReaderTuple), std::get<InIdx>(tempInputSpanStorage))...,
+                    refToSpan(std::get<OutIdx>(outputReaderTuple), std::get<OutIdx>(tempOutputSpanStorage))...
+            );
+        }(std::make_index_sequence<std::tuple_size_v<std::remove_cvref_t<decltype(inputReaderTuple)>>>(),
+               std::make_index_sequence<std::tuple_size_v<std::remove_cvref_t<decltype(outputReaderTuple)>>>());
     }
```

Comment:
I think this might be PMO and the original version is better/clearer. I'd propose to tackle this once the `-ftime-trace`, asm, or benchmark actually shows a benefit.